### PR TITLE
Make built-in types implement the new DigestSigner and DigestVerify interface

### LINKED ISSUE #120: Cannot sign/verify digests anymore (think Merkle tree, ledger, detached payload)
The implementation which allowed to sign the digests was removed in #100 in favor of passing raw content.

There are some cases when the user does not have content:
* for verification - only the digest is obtained from the Merkle tree, the user does not see the content
* for signing of sensitive content when only the digest is known, eg client constructs `toBeSigned`, creates a hash and sends to a service
* for cose receipts (eg [scitt receipts](https://datatracker.ietf.org/doc/draft-birkholz-scitt-receipts/)) where the structure `toBeSigned` is slightly different to the cose one

Is there a chance to extend the interfaces `Signer` and `Verifier` to allow digests again, similar to [`ecdsa.Sign`](https://pkg.go.dev/crypto/ecdsa#Sign) and [`ecdsa.VerifyASN1`](https://pkg.go.dev/crypto/ecdsa#VerifyASN1)?

### PR DESCRIPTION
**Pull request by qmuntal**
Fixes #120

### CODE REVIEW

File: signer.go [hunk 1]
```diff
@@ -23,6 +23,16 @@ type Signer interface {
 	Sign(rand io.Reader, content []byte) ([]byte, error)
 }
 
+// DigestSigner is an interface for private keys to sign digested COSE signatures.
```
Comment (OR13) [inline signer.go:26]: this is interface also an integration point with remote KMS? 

any time you have hardware isolated keys, you need a signer interface to request signatures from them.

**Review by OR13 — APPROVED** [review]
looks good to me, but better to get a review from a better go dev

**Comment by yogeshbdeshpande** [comment]
@qmuntal as mentioned in the meeting, please add the README explaining the new API

**Comment by qmuntal** [comment]
> @qmuntal as mentioned in the meeting, please add the README explaining the new API

Done.

**Comment by SteveLasker** [comment]
@yogeshbdeshpande, @shizhMSFT, @ivarprudnikov, @setrofim can you PTAL

File: rsa.go [hunk 2]
```diff
@@ -54,12 +60,20 @@ func (rv *rsaVerifier) Algorithm() Algorithm {
...
-	if err := rsa.VerifyPSS(rv.key, hash, digest, signature, &rsa.PSSOptions{
+	return rv.VerifyDigest(digest, signature)
+}
+
+// VerifyDigest verifies message digest with the public key, returning nil
+// for success.
+// Otherwise, it returns ErrVerification.
+//
+// Reference: https://datatracker.ietf.org/doc/html/rfc8152#section-8.1
+func (rv *rsaVerifier) VerifyDigest(digest, signature []byte) error {
```
Comment (ivarprudnikov) [inline rsa.go:75]: add digest type to be explicit, it should be `[]byte`

**Comment by ivarprudnikov** [comment]
Unit tests would be necessary to prove it works.

1. Create a keypair (both for RSA and ECDSA), then for each
2. Create a signer, cast it to DigestSigner and sign some "foobar" digest
3. Create a verifier, cast it to DigestVerifier and verify the signature along with the digest

Otherwise it looks OK

**Comment by OR13** [comment]
You might want your test strings to include some nasty unicode, unless you use some standard test strings from some the RFCs

**Review by setrofim — APPROVED** [review]
LGTM

File: verifier.go [hunk 3]
```diff
@@ -22,9 +22,21 @@ type Verifier interface {
 	Verify(content, signature []byte) error
 }
 
+// DigestVerifier is an interface for public keys to verify digested COSE signatures.
+type DigestVerifier interface {
+	Verifier
```
Comment (shizhMSFT) [inline verifier.go:27]: Does `DigestVerifier` require `Verify(content, signature []byte) error`?

If no, it should be replaced with
```suggestion
	// Algorithm returns the signing algorithm associated with the public key.
	Algorithm() Algorithm
```

File: signer.go [hunk 4]
```diff
@@ -23,6 +23,16 @@ type Signer interface {
...
+type DigestSigner interface {
+	Signer
```
Comment (shizhMSFT) [inline signer.go:28]: Does `DigestSigner` require `Sign(rand io.Reader, content []byte) ([]byte, error)`?

If no, it should be replaced with
```suggestion
	// Algorithm returns the signing algorithm associated with the public key.
	Algorithm() Algorithm
```

File: rsa.go (hunk 2 above)
Comment (qmuntal) [reply rsa.go]: Done

File: verifier.go (hunk 3 above)
Comment (qmuntal) [reply verifier.go:27]: Not really, good catch.

File: signer.go (hunk 4 above)
Comment (qmuntal) [reply signer.go:28]: Not really, good catch.

**Comment by qmuntal** [comment]
> Unit tests would be necessary to prove it works.
> 
> 1. Create a keypair (both for RSA and ECDSA), then for each
> 2. Create a signer, cast it to DigestSigner and sign some "foobar" digest
> 3. Create a verifier, cast it to DigestVerifier and verify the signature along with the digest
> 
> Otherwise it looks OK

> You might want your test strings to include some nasty unicode, unless you use some standard test strings from some the RFCs

@ivarprudnikov @OR13 done. I initially didn't add any test case because the built-in signers/verifiers are internally implemented in terms of `DigestSigner` and `DigestVerifier`, therefore the existing test coverage already applies to the new interfaces. I've added some basic tests, as suggested, to be more explicit.

File: README.md [hunk 5]
```diff
@@ -137,6 +137,16 @@ See [example_test.go](./example_test.go) for more examples.
...
 
+#### Signing and Verification of digested payloads
+
+When `cose.NewSigner` is used with PS{256,384,512} or ES{256,384,512}, the returned signer
+can be casted to the `cose.DigestSigner` interface, whose `SignDigest` method signs an
+already digested message.
+
+When `cose.NewVerifier` is used with PS{256,384,512} or ES{256,384,512}, the returned verifier
+can be casted to the `cose.DigestVerifier` interface, whose `VerifyDigest` method verifies an
+already digested message.
```
Comment (yogeshbdeshpande) [inline README.md:148]: Please add a similar construct here as well exemplifying, as you have added above, i.e.

See [example_test.go](https://github.com/veraison/go-cose/blob/82a1648f4a359716c9ab21981c884d780893e9ba/example_test.go) for more examples.

File: signer.go [hunk 6]
```diff
@@ -23,6 +23,17 @@ type Signer interface {
 	Sign(rand io.Reader, content []byte) ([]byte, error)
 }
 
+// DigestSigner is an interface for private keys to sign digested COSE signatures.
+type DigestSigner interface {
```
Comment (yogeshbdeshpande) [inline signer.go:27]: Question: Does this interface support signing digest from any Digest Algorithm ? Meaning should `go-cose` not restrict the digest signing to ONLY a known set of digest algorithms? This interface relies on caller to just pass digest as a substitute to content[] ?

Comment (qmuntal) [reply signer.go:27]: > Question: Does this interface support signing digest from any Digest Algorithm ? Meaning should go-cose not restrict the digest signing to ONLY a known set of digest algorithms

How could we restrict that? Everyone is free to implement the interface as needed, Go doesn't allow intercepting interface implementations to do custom validations. Also, in this PR we don't use this interface anywhere, users would have to construct the `cose.Sign1Message` themselves, so it's an experts-only feature.

> This interface relies on caller to just pass digest as a substitute to content[] ?

Yes

File: README.md (hunk 5 above)
Comment (qmuntal) [reply README.md:148]: Done.

Comment (qmuntal) [reply README.md:148]: But only added an example to example_test.go. This being an advanced feature, I prefer not to pollute the README.md.

File: README.md [hunk 7]
```diff
@@ -137,6 +137,16 @@ See [example_test.go](./example_test.go) for more examples.
 Untagged COSE_Sign1 messages can be signed and verified as above, using
 `cose.UntaggedSign1Message` instead of `cose.Sign1Message`.
 
+#### Signing and Verification of digested payloads
```
Comment (yogeshbdeshpande) [inline README.md:140]:
```suggestion
#### Signing and Verification of payload digest
```

File: README.md (hunk 5 above)
Comment (yogeshbdeshpande) [inline README.md:148]:
```suggestion
already digested message.

Please refer to [example_test.go](./example_test.go) for the API usage
```

Comment (yogeshbdeshpande) [reply README.md:148]: I never said, we should pollute the README. I just said point the reader to examples as shown below, how to use the interface.

**Review by yogeshbdeshpande — APPROVED** [review]
LGTM!
Thanks, I suggested some minor nits, but in general look ok!

File: README.md (hunk 5 above)
Comment (qmuntal) [reply README.md:148]: Oh, I misunderstood your comment. Done!

**Pull request merged**